pkg/services/rpcsrv/server.go

```diff
@@ -1669,6 +1662,127 @@ func (s *Server) findStates(ps params.Params) (any, *neorpc.Error) {
 	return res, nil
 }
 
+// getStateRootFromParam retrieves state root hash from the provided parameter
+// (only util.Uint256 serialized representation is allowed) and checks whether
+// MPT states are supported for the old stateroot.
+func (s *Server) getStateRootFromParam(p *params.Param) (util.Uint256, *neorpc.Error) {
+	root, err := p.GetUint256()
+	if err != nil {
+		return util.Uint256{}, neorpc.WrapErrorWithData(neorpc.ErrInvalidParams, "invalid stateroot")
+	}
+	if s.chain.GetConfig().Ledger.KeepOnlyLatestState {
+		curr, err := s.chain.GetStateModule().GetStateRoot(s.chain.BlockHeight())
+		if err != nil {
+			return util.Uint256{}, neorpc.NewInternalServerError(fmt.Sprintf("failed to get current stateroot: %s", err))
+		}
+		if !curr.Root.Equals(root) {
+			return util.Uint256{}, neorpc.WrapErrorWithData(neorpc.ErrUnsupportedState, fmt.Sprintf("state-based methods are not supported for old states: %s", errKeepOnlyLatestState))
+		}
+	}
+	return root, nil
+}
+
+func (s *Server) findStorage(reqParams params.Params) (any, *neorpc.Error) {
+	id, prefix, start, take, respErr := s.getFindStorageParams(reqParams)
+	if respErr != nil {
+		return nil, respErr
+	}
+	var (
+		i   int
+		end = start + take
+		res = new(result.FindStorage)
+	)
+	s.chain.SeekStorage(id, prefix, func(k, v []byte) bool {
+		if i < start {
+			i++
+			return true
+		}
+		if i < end {
+			res.Results = append(res.Results, result.KeyValue{
+				Key:   append(prefix, k...), // Don't strip prefix, as it is done in C#.
+				Value: v,
+			})
+			i++
+			return true
+		}
+		res.Truncated = true
+		return false
+	})
+	res.Next = i
+	return res, nil
+}
+
+func (s *Server) findStorageHistoric(reqParams params.Params) (any, *neorpc.Error) {
+	root, respErr := s.getStateRootFromParam(reqParams.Value(0))
+	if respErr != nil {
+		return nil, respErr
+	}
+	if len(reqParams) < 2 {
+		return nil, neorpc.ErrInvalidParams
+	}
+	id, prefix, start, take, respErr := s.getFindStorageParams(reqParams[1:], root)
+	if respErr != nil {
+		return nil, respErr
+	}
+
+	var (
+		end  = start + take
+		res  = new(result.FindStorage)
+		pKey = makeStorageKey(id, prefix)
+	)
+	// @roman-khimov, retrieving only the necessary part of the contract storage
+	// requires an mpt Billet refactoring, we can do it in a separate issue, create?
+	kvs, err := s.chain.GetStateModule().FindStates(root, pKey, nil, end+1) // +1 to define result truncation
+	if err != nil && !errors.Is(err, mpt.ErrNotFound) {
+		return nil, neorpc.NewInternalServerError(fmt.Sprintf("failed to find state items: %s", err))
+	}
```

Comment:
Not sure we really needed, it's relevant for large contract storages and latest storage items pages.

Reply:
Take mainnet and some NEO contract. It has a huge set of accounts and you'll need a substantial chunk of memory to hold them, so it'd be easy to cause DoS.

Reply:
Not fixed yet, I'll create a separate commit for it tommorow. But this can be solved easily, I've added one more new API that is similar to storage Seek.

Reply:
Fixed in a separate (last) commit.